### PR TITLE
Adding instructions for pre-filling the disk cache

### PR DESCRIPTION
**Pull request by arschles**
**What is the problem I am trying to address?**

There has been discussion in the `#athens` channel in Slack about how to pre-fill the Athens cache, and there's definitely a need to do so. There are no current documentation for how to do it either.

**How is the fix applied?**

I wrote a new documentation page for how to pre-fill using [pacmod](https://github.com/plexsystems/pacmod)

**Mention the issue number it fixes or add the details of the changes if it doesn't have a specific issue.**

n/a

cc/ @jpreese because you wrote much (all?) of pacmod, and @doddi because you were discussing it on Slack. ~There was one other person whose slack name is `Brian`, who was also discussing this, but I don't have their GH name, unfortunately. ~ Update: I know their name now! Hello, and thanks for bringing this up @brian73286 :wave:!

### CODE REVIEW

**Comment by jpreese** [comment]
LGTM 👍 . Let me know if you run into any issues with `pacmod`, I'd be more than happy to get it fixed up for you if its going to get more attention.

File: docs/content/configuration/prefill-disk-cache.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+$ export MODULE_NAME="github.com/my/module"
+$ export VERSION="v1.0.0
+```
+
+>Note: make sure your `VERSION` variable starts with a `v`
+
+Next, navigate to the top-level directory of the module source code, and run `pacmod` like this:
+
+```console
+$ pacmod pack $MODULE_NAME $VERSION
```
[the remark this answers is not in the document]
Comment (arschles) [reply docs/content/configuration/prefill-disk-cache.md]: @jpreese got it! So would `pacmod pack $VERSION .` work? If so, then I could change the below `mv` commands too

**Comment by arschles** [comment]
@jpreese thanks 💯 for the comments. I made changes to use the updated `pacmod`. Can you let me know what you think when you get a chance?

File: docs/content/configuration/prefill-disk-cache.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpreese) [reply docs/content/configuration/prefill-disk-cache.md]: Yes. `pacmod pack v1.0.0 .` would find your `go.mod` file in the current directory and create a `v1.0.0.info` file and `v1.0.0.zip` file in the current directory as well.

File: docs/content/configuration/prefill-disk-cache.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+- `$VERSION.info` - metadata about the module
+
+The `source.zip` file has a specific directory structure and the `$VERSION.info` has a JSON structure, both of which you'll need to get right in order for Athens to serve up the right dependency formats that the Go toolchain will accept.
+
+>We don't recommend that you create these assets yourself. Instead, use [pacmod](https://github.com/plexsystems/pacmod)
+
+To install the `pacmod` tool, run `go install` like this:
+
+```console
+$ go install github.com/plexsystems/pacmod
```
Comment (jpreese) [inline docs/content/configuration/prefill-disk-cache.md]: I think we should use `go get` here. `go install` assumes the code is already on the end-users machine.

File: docs/content/configuration/prefill-disk-cache.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+>Note: make sure your `VERSION` variable starts with a `v`
+
+Next, navigate to the top-level directory of the module source code, and run `pacmod` like this:
+
+```console
+$ pacmod pack $VERSION .
+```
+
+Once this command is done, you'll notice three new files in the same same directory you ran the command from:
```
Comment (jpreese) [inline docs/content/configuration/prefill-disk-cache.md]: same x2

File: docs/content/configuration/prefill-disk-cache.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
+---
+title: Pre-filling the Disk Cache
+description: How to pre-fill the disk cache
+weight: 4
+---
+
+One of the popular features of Athens is that it can be run completely cut off to the internet. In this case, though, it can't reach out to an upstream (e.g. a VCS or another module proxy) to fetch modules that it doesn't have in storage. So, we need to manually fill up the disk partition that Athens uses with the dependencies that we need.
```
Comment (jpreese) [inline docs/content/configuration/prefill-disk-cache.md]: I think it makes more sense to say "cut off _from_ the internet" here.

File: docs/content/configuration/prefill-disk-cache.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+```console
+$ mkdir -p $STORAGE_ROOT/github.com/my/module/$VERSION
+```
+
+Finally, make sure that you're still in the module source repository root directory (the same as you were in when you ran the `pacmod` command), and move your three new files into the new directory you just created:
+
+```console
+$ mv go.mod $STORAGE_ROOT/github.com/my/module/$VERSION/go.mod
+$ mv $VERSION.info $STORAGE_ROOT/github.com/my/module/$VERSION/$VERSION.info
+$ mv $VERSION.zip $STORAGE_ROOT/github.com/my/module/$VERSION/source.zip
```
Comment (jpreese) [inline docs/content/configuration/prefill-disk-cache.md]: The `.zip` file for modules is the same name as the `.info` file ( $VERSION )

File: docs/content/configuration/prefill-disk-cache.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+One of the popular features of Athens is that it can be run completely cut off to the internet. In this case, though, it can't reach out to an upstream (e.g. a VCS or another module proxy) to fetch modules that it doesn't have in storage. So, we need to manually fill up the disk partition that Athens uses with the dependencies that we need.
+
+This document will guide you through packaging up a single module called `github.com/my/module`, and inserting it into the Athens disk storage.
+
+# First, get the tools
+
+You'll need to produce the following assets from module source code:
+
+- `source.zip` - just the Go source code, packaged in a zip file
```
Comment (jpreese) [inline docs/content/configuration/prefill-disk-cache.md]: I mentioned it below as well, but the downloaded `zip` when fetching a module is $VERSION as well (e.g. `v1.0.0.zip`)

**Comment by jpreese** [comment]
@arschles no problemo. I added some notes after giving it a read through.

**Comment by brian73286** [comment]
Commenting here to associated myself as the referenced `Brian` from slack.

File: docs/content/configuration/prefill-disk-cache.md (hunk 6 above)
Comment (arschles) [reply docs/content/configuration/prefill-disk-cache.md]: @jpreese what I meant to say here is that Athens will need a `source.zip` in its disk storage location. Not sure if that was clear?

**Comment by arschles** [comment]
@jpreese I made some changes to help clear up some of your comments. I wasn't sure about the comments you made [here](https://github.com/gomods/athens/pull/1414/files#r333175469), [here](https://github.com/gomods/athens/pull/1414/files#r333175945), and [here](https://github.com/gomods/athens/pull/1414/files#r333171843). What I was trying to say was that Athens needs a `source.zip` on disk, but `pacmod` produces a `$VERSION.zip`. I don't know if I was clear enough throughout the document. What do you think?

**Comment by arschles** [comment]
cc/ @klingtnet

File: docs/content/configuration/prefill-disk-cache.md (hunk 5 above)
Comment (arschles) [reply docs/content/configuration/prefill-disk-cache.md]: @jpreese I think what you said went over my head. What should I change here?

Comment (jpreese) [reply docs/content/configuration/prefill-disk-cache.md]: Nothing! I think we're good here. There was a lot of confusion around how Athens stores its `.zip` by naming it `source.zip` and how the Go modules toolchain stores its zips (`$VERSION.zip`). Though from what I recall Athens still handles it correctly. I have a comment around here somewhere when I dove into this a bit ago.. but I've forgotten 🤣

Comment (arschles) [reply docs/content/configuration/prefill-disk-cache.md]: ohhhhh, I got you 😄 

Thanks for clarifying!

**Comment by arschles** [comment]
@jpreese and @brian73286 - are you all ok with merging this?

**Comment by jpreese** [comment]
LGTM.. the only reservation I had was that @brian73286 seemed to have some issues after following these steps, but I was unable to recreate the issue (https://github.com/gomods/athens/issues/1416).

I think we're OK to merge as its already been open a while, and we can iterate from there.

**Comment by arschles** [comment]
Thanks @jpreese. I wasn't able to reproduce that one either, but let's keep it open and I'll follow up right after I merge this.

Also, there is some related discussion going on in https://github.com/gomods/athens/issues/1459, so I may come back to these documentation in the near future to modify them.

**Pull request merged**